### PR TITLE
Implemented specialized class for Euler angles

### PR DESCRIPTION
**Pull request by evbernardes**
#### Brief description of what is fixed or changed
This is a simple and incomplete implementation of a class dealing with Euler angles.

I'm making this PR to see if people would find it useful for me to keep working on!

Many properties, like power and composition, that cannot be usually applied to Euler angles are dealt with here by first converting it to a Quaternion.

Very inefficient! Which is why I put warnings everywhere. 

Moreover, special care is taken when mixing angles from different sequences.

Provided function to convert from one sequence to another.

#### Other comments
At the moment, most of what it does is to make some calls to the `Quaternion` class, making it easy to test.

This class stores information about the sequences, which can be convenient. Also implemented some printers for it:
```python
Euler(pi / 2, 0, pi / 4, seq='zyz')
```

Prints:
```math
\left[\begin{matrix}\left(\frac{\pi}{2}\right)\\ 0\\ \left(\frac{\pi}{4}\right)\end{matrix}\right]^{zyz}
```

This is maybe a bit ugly though, would love some better ideas.

One of the advantages, for example, is that this class always tries to see if the angles are in the set `[-pi, pi]` after `subs` and `eval`:
```python
# setting quaternion
>>> seq = 'zyz'
>>> values = (1, 0, -7, -1)
>>> q = Quaternion(a, b, c, d)

# setting substitutions
>>> subs = [(var, sympify(value)) for var, value in zip(q.args, values)]

# without using Euler, the subs gives a value smaller than -pi
>>> angles = Matrix(q.to_euler(seq)).subs(subs)
>>> print(angles.evalf())
Matrix([[-3.92699081698724], [2.74289823818859], [2.35619449019234]])

# using Euler, the subs gives a correct value
>>> angles= Euler(*q.to_euler(seq)).subs(subs).to_Matrix()
>>> print(angles.evalf())
Matrix([[2.35619449019234], [2.74289823818859], [2.35619449019234]])
```
As a next step, all of the Quaternion / Euler conversion functions could be migrated to `Euler`, and `Quaternion.to_euler` could be a call to a future `Euler.from_quaternion` method.

#### Release Notes

* algebras
  * Implemented specialized class for Euler angles.

### CODE REVIEW

File: sympy/algebras/euler.py [hunk 1]
```diff
@@ -0,0 +1,634 @@
...
+            raise ValueError("arguments have to be commutative")
+        else:
+            obj = Expr.__new__(cls, alpha, beta, gamma)
+            obj._alpha = alpha
+            obj._beta = beta
+            obj._gamma = gamma
+            obj._info = info
+            obj._q = None
+            obj._R = None
+            return obj
```
Comment (oscarbenjamin) [inline sympy/algebras/euler.py:109]: Basic subclasses should avoid creating lots of attributes that are not part of the args of the object.

This will not satisfy the func-args invariant: `obj == obj.func(*obj.args)` whenever `seq` is something other than `'ZYX'`.

Comment (evbernardes) [reply sympy/algebras/euler.py]: Added `seq` to args, but now for some reason I have a bug with `evalf`:
```python
    euler2.evalf()

  File /sympy/sympy/core/evalf.py:1644 in evalf
    result = evalf(self, prec + 4, options)

  File /sympy/sympy/core/evalf.py:1492 in evalf
    re, im = as_real_imag()

  File /sympy/sympy/core/expr.py:1931 in as_real_imag
    return (re(self), im(self))

  File /sympy/sympy/core/cache.py:72 in wrapper
    retval = cfunc(*args, **kwargs)

  File /sympy/sympy/core/function.py:470 in __new__
    result = super().__new__(cls, *args, **options)

  File /sympy/sympy/core/cache.py:72 in wrapper
    retval = cfunc(*args, **kwargs)

  File /sympy/sympy/core/function.py:310 in __new__
    evaluated = cls.eval(*args)

  File sympy/sympy/functions/elementary/complexes.py:92 in eval
    elif not term.has(I) and term.is_extended_real:

  File sympy/sympy/core/cache.py:72 in wrapper
    retval = cfunc(*args, **kwargs)

  File sympy/sympy/core/basic.py:1267 in has
    return self._has(iterargs, *patterns)

  File sympy/sympy/core/basic.py:1357 in _has
    for i in iterargs(self):  #

  File /sympy/sympy/core/traversal.py:31 in iterargs
    args.extend(i.args)

AttributeError: 'str' object has no attribute 'args'
```

I think I should sympify `seq`, but then I'd need to recover the string from the symbol. How can I do that?

File: sympy/__init__.py [hunk 2]
```diff
@@ -230,7 +230,7 @@ def __sympy_debug():
         differentiate_finite, periodicity, not_empty_in, AccumBounds,
         is_convex, stationary_points, minimum, maximum)
 
-from .algebras import Quaternion
+from .algebras import (Quaternion, Euler)
```
Comment (sylee957) [inline sympy/__init__.py:233]:
```suggestion
from .algebras import Quaternion, Euler
```

File: sympy/algebras/euler.py [hunk 3]
```diff
@@ -0,0 +1,636 @@
...
+
+    intrinsic = seq.isupper()
+    extrinsic = seq.islower()
+    if not (intrinsic or extrinsic):
+        raise ValueError("seq must either be fully uppercase (for extrinsic "
+                         "rotations), or fully lowercase, for intrinsic "
+                         "rotations).")
+
+    i, j, k = seq.lower()
+    if (i == j) or (j == k):
```
Comment (sylee957) [inline sympy/algebras/euler.py:27]:
```suggestion
    if i == j or j == k:
```

File: sympy/algebras/euler.py (hunk 1 above)
Comment (sylee957) [reply sympy/algebras/euler.py]: A basic workaround is to use `core.Symbol`, or subclass nullary `Basic` instance of each alphabets
(`class X(Basic): pass`, ...)

Comment (evbernardes) [reply sympy/algebras/euler.py]: Alright I think I managed to work around this!

File: sympy/algebras/euler.py [hunk 4]
```diff
@@ -0,0 +1,676 @@
...
+
+        if any(i.is_commutative is False for i in [alpha, beta, gamma]):
+            raise ValueError("arguments have to be commutative")
+        else:
+            obj = Expr.__new__(cls, alpha, beta, gamma, seq)
+            obj._alpha = alpha
+            obj._beta = beta
+            obj._gamma = gamma
+            obj._info = info
+            obj._q = None
```
Comment (oscarbenjamin) [inline sympy/algebras/euler.py:111]: These attributes should not be set. Instead everything should be derived from `self.args`.

Comment (evbernardes) [reply sympy/algebras/euler.py]: Alright, did just that, and removed the `info` dict I had defined.

I left the `_q` attribute to lazily calculate the quaternion since it might give some speed advantages.

Moreover, `Euler.seq` gives `str(Euler.args[3])` instead of `Euler.args[3]`, this is because of the workaround I had to do to put `seq` as a `sympify`ed `Symbol` in `args`.

File: sympy/algebras/euler.py [hunk 5]
```diff
@@ -0,0 +1,667 @@
...
+        """Get all angles as a tuple"""
+        return self.args[:3]
+
+    def __getitem__(self, key):
+        """Get all angles by index"""
+        return self.angles[key]
+
+    @property
+    def info(self):
+        return self._info
```
Comment (oscarbenjamin) [inline sympy/algebras/euler.py:112]: This will fail now

Comment (evbernardes) [reply sympy/algebras/euler.py]: Oops, indeed!

File: sympy/angles/euler.py [hunk 6]
```diff
@@ -0,0 +1,688 @@
...
+        angles = q.to_euler(seq,
+                            angle_addition=angle_addition,
+                            avoid_square_root=avoid_square_root)
+        return Euler(*angles, seq=seq)
+
+    def add(self, other):
+        """Adds Euler angles elements element-wise, treating it as a Matrix.
+
+        This does not have much mathematical meaning, are you sure this is what
+        you want?
```
Comment (oscarbenjamin) [inline sympy/angles/euler.py:289]: This does seem a bit odd

File: sympy/angles/euler.py [hunk 7]
```diff
@@ -0,0 +1,688 @@
...
+        Euler(pi/2,  0,  0, ZYX)
+
+        """
+        q = Quaternion.from_rotation_matrix(M)
+        q = trigsimp(q)
+        return Euler.from_quaternion(q, seq)
+
+    @staticmethod
+    def _generic_mul(lhs, rhs):
+        """Generic multiplication.
```
Comment (oscarbenjamin) [inline sympy/angles/euler.py:444]: If adding doesn't really make sense then what does multiplication mean?

File: sympy/angles/euler.py [hunk 8]
```diff
@@ -0,0 +1,688 @@
...
+        >>> euler.mul(euler).evalf()
+        Euler(1.5707963267949,  0,  0, ZYX)
+        """
+        return self._generic_mul(self, _sympify(other))
+
+    def __add__(self, other):
+        return self.add(other)
+
+    def __radd__(self, other):
+        return self.add(other)
```
Comment (oscarbenjamin) [inline sympy/angles/euler.py:603]: It is usually better to avoid defining these methods in Expr subclasses. I'm not sure if there is a better way though.

File: sympy/angles/euler.py [hunk 9]
```diff
@@ -0,0 +1,688 @@
...
+        if p == -1:
+            return self.inverse()
+
+        q = self.to_quaternion() ** p
+        return Euler.from_quaternion(q, seq=self.seq)
+
+    @property
+    def is_zero_rotation(self):
+        """Returns true if the angles are all zero"""
+        return all(i.is_zero for i in self.angles)
```
Comment (oscarbenjamin) [inline sympy/angles/euler.py:675]: This should use `fuzzy_and` rather than `all` so that it handles `is_zero = None`.

File: sympy/angles/euler.py (hunk 7 above)
Comment (evbernardes) [reply sympy/angles/euler.py:444]: I thought of implementing multiplication as a composition of rotations, just like for a rotation matrix and for a quaternion.

Except, of course, that most of the time Euler angles are horrible for composing, this is why I implemented it as a multiplication of quaternions followed by a conversion to Euler angles.

File: sympy/angles/euler.py (hunk 6 above)
Comment (evbernardes) [reply sympy/angles/euler.py:289]: The sum of two rotation matrices isn't a rotation matrix, the same way that a sum of a rotation quaternion is not a quaternion.

I was not sure if it was a good idea, but I decided to implement the sum of Euler angles in the same vein: adding all of the elements, but not necessarily giving back something as meaningful as the other values.

I can totally remove all of the definitions of additions though.

File: sympy/angles/euler.py (hunk 9 above)
Comment (evbernardes) [reply sympy/angles/euler.py:675]: An right, didn't know that, thanks!

File: sympy/angles/euler.py (hunk 6 above)
Comment (sylee957) [reply sympy/angles/euler.py:289]: I think that multiplication can also be left undefined for the cases where it must need coversion to matrix or quaternion.

Comment (evbernardes) [reply sympy/angles/euler.py:289]: I think that composition with other angles, although inefficient, could be a good addition to this class that would at least work well for numerical values or simple cases.

What if I separate it in a subfunction called `Euler.compose` and raise a `ValueError` inside of  `__mul__` when two Euler values are being multiplied, giving out a message that says `Multiplication undefined between two Euler instances. For rotation composition, use the 'compose' method.`?

Comment (oscarbenjamin) [reply sympy/angles/euler.py:289]: Why define these methods at all? Are they actually needed for something?

Comment (evbernardes) [reply sympy/angles/euler.py:289]: You mean the addition/multiplication as matrices, or the composition?

Comment (evbernardes) [reply sympy/angles/euler.py:289]: Maybe the least ambiguous implementation would be for `__mul__` to be implemented as a call to `compose` only when both elements are instances of `Euler`.

For other inputs of `__mul__`, and any input of  `__add__` or `__sub__`, the usual type error with a message saying that for treating the elements as a `Matrix`, the user should use `to_Matrix()`.

And `__pow__` could keep being the composition with itself as already implemented.

Comment (oscarbenjamin) [reply sympy/angles/euler.py:289]: I think if there's any ambiguity about what operations like `+` and `*` mean then it is better not to define them at all. If the operations are just going to convert to matrix or quaternion then it would be better for the user to use matrices or quaternions.

What is this class going to be used for? Don't just add methods for the sake of it.

It isn't clear to me that this class should be added at all.

Comment (evbernardes) [reply sympy/angles/euler.py:289]: This implementation is a rather simple one for now, because I wanted to first see if there was any interest in such a class.

The idea with it is to further on implement some helpers for dealing with Euler angles. For example, one thing that I find particularly confusing but is used a LOT in research papers is to use Euler angles when deriving the Equations of Motion of a body, and then using Euler angles again to derive a control law for them.

With quaternions, this is easy, since, for an angular velocity $\pmb \omega$:
```math
\dot q = \frac{1}{2} q \odot \begin{bmatrix} 0 \\ \pmb \omega \end{bmatrix} 
```

With Euler angles, having an expression of their derivaties is a LOT more complicated, because for three angles $\alpha$, $\beta$ and $\gamma$, what we have is:
```math
\begin{align}
\dot R &= \pmb A \begin{bmatrix} \dot \alpha \\ \dot \beta \\ \dot \gamma \end{bmatrix} \\
\dot R &= R \hat{\pmb \omega}
\end{align}
```

Where this matrix $\pmb A$ is always a complicated expression that depends on the sequence used.

One thing I have planned is to implement a helper for calculating these matrices for example, to help me navigate these research papers a bit better.

Since I'm probably not the only one having to deal with this, then I imagined I'd first see if such an Euler angle helper class could be of any interest in Sympy.

**Comment by oscarbenjamin** [comment]
I'm going to respond down here rather than in a thread commenting on a line of code above. https://github.com/sympy/sympy/pull/24523#discussion_r1071233579

It's unclear to me why it is beneficial to have a class for Euler angles rather than just representing them as a matrix or a tuple or something. That is not to say that this class is a bad idea but I don't really understand why the particular combination of methods have been added to this class or what its purpose is.

For example we already have `Quaternion.to_euler` so do we need `Euler.from_quaternion` (which is just a wrapper for the former)?

Some methods like `+` and `*` don't really seem to make sense and just seem to be added because once you start defining a class you start thinking about things like that.

When I strip it back the only method on this class that is not necessarily redundant for the functionality that already exists in the Quaternion class seems to be the `to_rotation_matrix` method. That could just be a function though (i.e. we don't need a class just for one function).

**Comment by evbernardes** [comment]
> It's unclear to me why it is beneficial to have a class for Euler angles rather than just representing them as a matrix or a tuple or something. That is not to say that this class is a bad idea but I don't really understand why the particular combination of methods have been added to this class or what its purpose is.

The basic difference is that I wanted have some sort of `Matrix` that also stores the `sequence` and is always making sure the angles are being treated in the correct sequence. 

> For example we already have `Quaternion.to_euler` so do we need `Euler.from_quaternion` (which is just a wrapper for the former)?

And same for this case. I have worked for example with a multibody system in which it made sense to use one sequence for one body and a different sequence on another body...

> Some methods like `+` and `*` don't really seem to make sense and just seem to be added because once you start defining a class you start thinking about things like that.

Completely agree for `+`, what I like about keeping `*` though is that since they are calls for compose, I find it easier to translate, for example, a rotation matrix composition to a rotation Angles composition in the correct order.

> When I strip it back the only method on this class that is not necessarily redundant for the functionality that already exists in the Quaternion class seems to be the `to_rotation_matrix` method. That could just be a function though (i.e. we don't need a class just for one function).

I was indeed not certain if such a Class would have it's place in `SymPy`. But as I have seen `EulerAngles` classes implemented in other big numerical libraries, like `Eigen`, for example, in which they are usually only an array + sequence, I thought it could have more sense in an algebra system like SymPy.

Taking the example of `Eigen`, their conversion algorithm gives a different set of Euler angles which is pretty unusual, but also valid. In this implementation here, I also added the method `switch_set` to give an equivalent set of angles that results in the same final rotation.

Basically, I think there are many possible little helpers for Euler angles that would make it worth it to have a specialized Class. While this could all be implemented by the user, I'd argue that one advantage of having it coded as a module is that it would already implement all of the necessary tests.  

If you prefer, I can close this PR now. I'll probably keep working on this as a side-project for now since I think this will be at least pretty useful for me.

Then (if you'd agree), I can try to make a new PR once it has all of the features I wanted.

**Comment by oscarbenjamin** [comment]
> The basic difference is that I wanted have some sort of `Matrix` that also stores the `sequence` and is always making sure the angles are being treated in the correct sequence.

I think that this *maybe* makes sense. Adding a symbolic class is a significant overhead though because it means there are a lot of methods that will have to be resolved somehow.

> I'd argue that one advantage of having it coded as a module is that it would already implement all of the necessary tests.

The disadvantage is the need to maintain this code forever. Basically there is always a cost in adding new code.

By the way this should definitely not be a top-level `sympy.angles` module. This is way too niche to be at the top-level namespace. Probably it should just all be in the quaternion module or perhaps the vector module.

File: sympy/angles/euler.py [hunk 10]
```diff
@@ -0,0 +1,688 @@
...
+
+
+def _wrap(angle):
+    if angle.is_number:
+        return (angle + pi) % (2 * pi) - pi
+    else:
+        return angle
+
+
+class Euler(Expr):
```
Comment (oscarbenjamin) [inline sympy/angles/euler.py:49]: This should not subclass Expr. Expr means things that can be added, multiplied etc and can be in the args of an Add or Mul. This class does not make sense in any of those contexts.

**Comment by oscarbenjamin** [comment]
> > The basic difference is that I wanted have some sort of `Matrix` that also stores the `sequence` and is always making sure the angles are being treated in the correct sequence.
> 
> I think that this _maybe_ makes sense.

If this is the purpose then most of the methods should probably be removed though. I don't think it's useful to add lots of methods that are just shorthand for converting to a quaternion and then doing some operation. Likewise a lot of the attributes seem redundant. It is better to keep things simple.

**Comment by evbernardes** [comment]
> The disadvantage is the need to maintain this code forever. Basically there is always a cost in adding new code.

Ok, understandable!
 
> By the way this should definitely not be a top-level `sympy.angles` module. This is way too niche to be at the top-level namespace. Probably it should just all be in the quaternion module or perhaps the vector module.

I originally added it inside of the `algebras` module together with `quaternion`, but removed it since it's not really an algebra...

Should I get it back anyways? And should the doc page be a separate page?

File: sympy/angles/euler.py (hunk 10 above)
Comment (evbernardes) [reply sympy/angles/euler.py:49]: Alright, what should I use as a mother class?

**Comment by evbernardes** [comment]
> If this is the purpose then most of the methods should probably be removed though. I don't think it's useful to add lots of methods that are just shorthand for converting to a quaternion and then doing some operation. Likewise a lot of the attributes seem redundant. It is better to keep things simple.

Alright, I see. In this case I think I'm closing the PR instead and I'll keep it as a separate private project. Probably makes more sense overall.

File: sympy/angles/euler.py (hunk 10 above)
Comment (oscarbenjamin) [reply sympy/angles/euler.py:49]: It should probably be Basic.

**Comment by sylee957** [comment]
> But as I have seen EulerAngles classes implemented in other big numerical libraries, like Eigen, for example, in which they are usually only an array + sequence, I thought it could have more sense in an algebra system like SymPy.

The difference in 'Eigen' is that they have completely different subclassing structure.
I think that sympy could been better in OOP if things were designed in SOLID principle, but it is nowhere near it, so that's why I'm also warning about sympy's OOP interface.
But I assume it is about about lacking support in formal verification tools that forces people follow strict programming rules.

**Comment by evbernardes** [comment]
> The difference in 'Eigen' is that they have completely different subclassing structure, that make their implementation easier, but not easier in sympy. I think that sympy could been better in OOP if things were designed in SOLID principle, but it is nowhere near it, so that's why I'm also warning about sympy's OOP interface. But I assume it is about about lacking support in formal verification tools that forces people follow strict programming rules.

Fair point!

An EulerAngle specialized class still makes sense for me, for some theoretical analysis I had to do, so I'll implement it separately when I have the time, including the derivative matrix for linearization.

Then I might open an issue or discussion once everything I plan on implementing is there.
